lldb/include/lldb/Core/Debugger.h

```diff
@@ -731,8 +747,11 @@ class Debugger : public std::enable_shared_from_this<Debugger>,
   lldb::TargetSP m_dummy_target_sp;
   Diagnostics::CallbackID m_diagnostics_callback_id;
 
-  lldb_private::DebuggerDestroyCallback m_destroy_callback = nullptr;
-  void *m_destroy_callback_baton = nullptr;
+  std::recursive_mutex m_destroy_callback_mutex;
```

Comment:
Does this actually **need** to be thread safe? APIs being thread safe is of course a good thing, but in its current state, there are plenty of methods in the Debugger class that are not guaranteed to be thread safe. Unless there's a need for this to be safe, I wouldn't bother with the new mutex... 

Either way this shouldn't be a _recursive_ mutex.

Reply:
Handling a callback could add or remove other callbacks, it does need to be thread safe. And as much as I dislike recursive mutex as well, is there another way you could handle those scenarios safely?

Reply:
> Handling a callback could add or remove other callbacks
IIUC, this answers the "recursive" part of @JDevlieghere 's comment.

> Does this actually need to be thread safe?
I had the same question in my mind when the thread safety requirement was raised. Now I look at it, the only other mutex in the same class is the global debugger list, which does sound important enough to not be messed up.

IMHO, the answer here depends on whether or not we already have a clear future where the class need to be thread safe.  If such future hasn't been aligned, I feel it can be unsafe for now, and we can add thread-safety as one-go once such direction has been aligned. WDYT?

Reply:
@bulbazord But that's only if you call `Debugger::Destroy` (and therefore run the callbacks) on another thread, right? Given nothing else in `(SB)Debugger` provides such safety guarantees (besides the global debugger list) I'd say let's not overthink this and if we want to change that, let's do it holistically for the whole class instead of piecemeal.

Reply:
@JDevlieghere Yes, I don't think we need to synchronize across threads, that's a good point. I was thinking about what if any of the callbacks add, remove, or otherwise invoke other callbacks. But that shouldn't require a mutex.

@royitaqi I think you can remove the mutex :)

Reply:
That's true of running the destroy chain callbacks, but you certainly can add a callback from more than one thread at a time.  I think you need to at least protect adding.

Reply:
BTW:
> I think you need to at least protect adding.

To me, it's either none, or all.

If we decide to protect adding (multiple adds from different threads not during destroy), then we should just protect all of add/remove/destroy, because any of these operations can happen at the same time from multiple threads.

FWIW I agree with what @JDevlieghere said:

> I'd say let's not overthink this and if we want to change that, let's do it holistically for the whole class instead of piecemeal.